js/witness_calculator.js

```diff
@@ -160,7 +159,45 @@ export default async function builder(code, options) {
                 }
             }
         }
-    });
+    };
+
+    WebAssembly.instantiate(wasmModule, importsObject);
+
+    function getMessage() {
+        var message = "";
+        var c = instance.exports.getMessageChar();
+        while ( c != 0 ) {
+            message += String.fromCharCode(c);
+            c = instance.exports.getMessageChar();
+        }
+        return message;
+    }
+
+    function p2str(p) {
+        const i8 = new Uint8Array(memory.buffer);
+
+        const bytes = [];
+
+        for (let i=0; i8[p+i]>0; i++)  bytes.push(i8[p+i]);
+
+        return String.fromCharCode.apply(null, bytes);
+    }
+
+};
+
+export default async function builder(codeOrWasmInstance, options) {
+
+    options = options || {};
+
+    let instance;
+
+    // In the event that you are running witness generation multiple times, it's
+    // better to reuse an existing wasm instance.
+    if (codeOrWasmInstance instanceof WebAssembly.Instance) {
+        instance = codeOrWasmInstance;
+    } else {
+        instance = await createWASMInstance(codeOrWasm, options);
+    }
 
     if (typeof instance.exports.getVersion == 'function') {
```

Comment:
These checks should be inside createWASMInstance and version object returned together with the instance (new wrapper object for wasm instance + version is needed)

Reply:
Also variable `wc` is not declared in this function, but is assigned later in the code AND is used inside createWASMInstance. So maybe it would be easier to move all the rest of the code including instantiation of WitnessCalculatorCircom2/1 to createWASMInstance too (and rename the function to createWitnessCalculatorInstance)

Reply:
you're totally right, I must have been really sleep deprived when I reworked this :man_facepalming: 

really the only change that's required here is 

```js
export default async function builder(code, options) {
    let instance;
    let wc;

    // Only circom 2 implements version lookup through exports in the WASM
    // We default to `1` and update if we see the `getVersion` export (major version)
    // These are updated after the instance is instantiated, assuming the functions are available
    let majorVersion = 1;
    // After Circom 2.0.7, Blaine added exported functions for getting minor and patch versions
    let minorVersion = 0;
    // If we can't lookup the patch version, assume the lowest
    let patchVersion = 0;

    if (code instanceof WebAssembly.Instance) {
        instance = code;
    } else {
        instance = //same definition as before
    }
    
    //same as before
```

The new diff looks bad because of new indentation, but this is now the only change